### DISCUSSION TITLE
How is interrupts handled in multi-core?

### DISCUSSION THREAD

**Discussion by drakejest** [earlephilhower/arduino-pico#1352] [Q&A]
Im assigning core 2 to handle data acquisition to the sensors, these involves multiple interrupts and i was wonderinng how the arduino pico core  handles these. ( I dont see an interrupt section in the docs)

If i do an attachInterrupt() on loop2/setup2 i assume it will only interrupt core 2 right? Core 1 will remain unaffected?

Does that also mean noInterrupts() and Interrupts() will only apply to the specific core they are called upon?

**Comment by earlephilhower** [comment]
It's core0 and core1, not core1 and core2, so be sure to use `setup1()` and `loop1()` not `setup2()`. :)

Multicore you should read the RP2040 datasheet and be aware that Arduino functions/libraries are generally not multicore or multithread safe.

At a high level, each core has its own controller so only the core you enable GPIO interrupts on will take the IRQ.  Only `attachInterrupt` on one core or you'll have some very odd behavior.  Same for calling things like LittleFS/etc.  `Serial` is special-cased and can be called from both cores and things generally "just work"  but that's the exception, not the rule.

`noInterrupts` etc. only affect the core it is run on since there are separate interrupt controllers.  Look at the `rp2040.idleOtherCore` docs if you need to freeze the other core.  Also look at the PicoSDK mutexes if you need to synchronize things between cores.

It's not *really* that hard, but it does take care to use both cores successfully.  General multicore/multithread rules apply here.

**Reply by drakejest** [reply]
Yeah arduino API is not really designed for multi-core use. 

Can you please elaborate more on _only do `attachInterrupt` on one core_. Because i would like to do attachInterrupt on both cores. theoretically they dont interfere with each other since their job is only to time stamp, they also have their own global variable, and those variables are only going to be used by their own respective cores

**Reply by earlephilhower** [reply]
There is only one GPIO block and only 1 interrupt for *all* pins, not individual IRQs for each pin.  So you end up with a race condition because depending on which core services the IRQ first.  Or, worse, you get both taking the interrupt at the same time and trying to write to a single GPIO control register at the same time.

**Reply by drakejest** [reply]
so there is only 1 IRQ for both cores? If you are only reading data from GPIO there should be no problem with race condition as you are not writing to any register. Its more of if pin 5 triggers the interrupt core0 should stop what its doing and process the data, core1 continue as normal. if pin 3 triggers the interrupt core0 continue as normal, core1 should stop and process the data.

**Reply by earlephilhower** [reply]
You need to clear the IRQ when you're done or it'll never re-trigger, so that's the crux of the issue.  And who/when it's cleared is the problem.

It's really not going to work well, but feel free to give it a try.  You might be able to hack the core to make something like that happen, but I would recommend thinking of another way to get what you want done.  Service all IRQs from one block on one core and then, maybe, generate a different IRQ (i.e. the intercore FIFO) to do something elsewhere if needed.

**Reply by drakejest** [reply]
Speaking of race conditions and only one GPIO register, wont there be a problem just by doing digitalWrite() on both cores?

**Reply by drakejest** [reply]
_You need to clear the IRQ when you're done or it'll never re-trigger, so that's the crux of the issue. And who/when it's cleared is the problem._

Ahh this clears things up, this is the part that i did not know needs to be done. So yes there is no way around it, has to be done on one core only.

**Reply by mweetman2** [reply]
I have a similar issue. I have a timer interrupt that I want to service on core1.
Using the RPI_PICO_TimerInterrupt library, I attach the timer interrupt in setup1(), so I assumed this would then run the interrupt handler on core1. However, the interrupt handler appears to run on core0.
I suspect  RPI_PICO_TimerInterrupt wasn't written for multicore? 
Still, I should be able to program the core1 NVIC to enable the timer interrupt, but would also need to add the TimerInterrupt vector.
Any help appreciated, THANKS.

**Reply by mweetman2** [reply]
No worries. I ditched RPI_PICO_TimerInterrupt for an alarm pool, where I can specify the core running the interrupt handler. 
Great tips in this post.
Thanks
